internal/scp/scp.go

```diff
@@ -0,0 +1,235 @@
+func (c *Config) Setup() error {
+	// This wouldn't be nil in testing situations but in
+	// general it will be nil for most runtime scenarios.
+	// This allows us to mock the filesystem in testing
+	// but for live running of the plugin we'll use the
+	// real file system.
+	if c.fs == nil {
+		c.fs = afero.NewOsFs()
+	}
+
+	// Pickup the scp & sshpass binaries from whatever location
+	// they might be currently installed into. Inside a plugin this
+	// should stay static, but when debugging and running this
+	// outside of the container it's nice if it picks up user binaries.
+	for _, path := range openssh.BinSearchLocations {
+		tempSCPPath := fmt.Sprintf("%s/scp", path)
+		tempSSHPath := fmt.Sprintf("%s/ssh", path)
+		tempSSHPassPath := fmt.Sprintf("%s/sshpass", path)
+		if ok, _ := afero.Exists(c.fs, tempSCPPath); ok && len(c.locationSCPbinary) == 0 {
+			c.locationSCPbinary = tempSCPPath
+		}
+		if ok, _ := afero.Exists(c.fs, tempSSHPath); ok && len(c.locationSSHbinary) == 0 {
+			c.locationSSHbinary = tempSSHPath
+		}
+		if ok, _ := afero.Exists(c.fs, tempSSHPassPath); ok && len(c.locationSSHPASSbinary) == 0 {
+			c.locationSSHPASSbinary = tempSSHPassPath
+		}
+	}
+
+	if c.locationSCPbinary == "" {
+		return openssh.ErrMissingSCP
+	}
+
+	if c.locationSSHbinary == "" {
+		return openssh.ErrMissingSSH
+	}
+
+	if c.locationSSHPASSbinary == "" {
+		return openssh.ErrMissingSSHPASS
+	}
+
+	if c.IdentityFileContents != "" {
+		filename, err := openssh.CreateRestrictedFile(c.fs, openssh.TempIdentityFilePrefix, c.IdentityFileContents)
+		if err != nil {
+			return err
+		}
+		c.IdentityFilePath = append([]string{filename}, c.IdentityFilePath...)
+	}
+
+	if c.SSHPassword != "" {
+		filename, err := openssh.CreateRestrictedFile(c.fs, openssh.TempPasswordPrefix, c.SSHPassword)
+		if err != nil {
+			return err
+		}
+		c.locationPasswordFile = filename
+	}
+
+	if c.SSHPassphrase != "" {
+		filename, err := openssh.CreateRestrictedFile(c.fs, openssh.TempPassphrasePrefix, c.SSHPassphrase)
+		if err != nil {
+			return err
+		}
+		c.locationPassphraseFile = filename
+	}
+
+	return nil
+}
+
+func (c *Config) Binary() string {
+	if c.useSSHPass() {
+		return c.locationSSHPASSbinary
+	}
+
+	return c.locationSCPbinary
+}
+
+func (c *Config) Arguments() []string {
+	args := []string{}
+
+	// sshpass expects to be first in the chain of commands called
+	// so if we're using it, we'll need to bump all arguments to the end
+	// and set any sshpass flags by the user before specifying the scp binary.
+	if c.useSSHPass() {
+		if len(c.SSHPASSFlags) == 0 {
+			args = append([]string{c.locationSSHPASSbinary}, openssh.DefaultSSHPassFlags...)
+		} else {
+			args = append([]string{c.locationSSHPASSbinary}, c.SSHPASSFlags...)
+		}
+
+		if len(c.SSHPassword) > 0 {
+			args = append(args, "-f")
+			args = append(args, c.locationPasswordFile)
+		} else if len(c.SSHPassphrase) > 0 {
+			args = append(args, "-Ppassphrase")
```

Comment:
If I'm interpreting the question correctly, it's the ask about why do we need `"-Ppassphrase"` and it's because the sshpass binary doesn't know the difference between passwords or passphrases. It only looks for output from the respective binary to see if it contains the keyword. The default thing it searches for is `assword:` to match case indefinite Password and password. When we're using a passphrase instead we have to tell the binary to search on a different keyword, which is `passphrase` in this case as I don't think it can appear capitalized but could change to `assphrase` to be sure.

Reply:
ok, thanks for the explanation. i did check the man pages and it didn't clear that up much at all for me.